api/src/main/java/org/eclipse/microprofile/config/ConfigValueResolver.java

```diff
@@ -0,0 +1,294 @@
+ * {@link #throwOnFailedConversion()}.
+ *
+ * <p>
+ * Usually conversion relies on registered {@link org.eclipse.microprofile.config.spi.Converter}s.
+ * Ad-hoc conversion can be done using {@link #asConvertedBy(Function, Object)}.
+ * Values resolved as {@link String} are not converted.
+ * Values resolved as {@code String[]} are only split but elements are not converted.
+ *
+ * <p>
+ * Typed defaults are provided with one of the {@code as}-methods.
+ * Raw property string defaults can be provided using {@link #withDefault(String)}.
+ * As the API can not capture if such a raw default has been provided it still requires to provide a typed default
+ * or use {@link Optional} to ensure resolution will return a value.
+ * If a raw {@link String} is provided it takes precedence over a typed default.
+ * Should conversion of a raw default fail the typed default is used
+ * or in case of {@link Optional} the property is considered not present.
+ *
+ * <p>
+ * The only way {@code as}-methods might return a {@code null} reference is by using {@code null} as typed default.
+ * In such case it is considered the intention of the caller and therefore not problematic.
+ *
+ * @author <a href="mailto:jan.bernitt@payara.fish">Jan Bernitt</a>
+ */
+public interface ConfigValueResolver {
+
+    /**
+     * Adds a profile to property resolution which is applied when requesting a value.
+     * Multiple calls to this method on the same instance override prior profile values.
+     *
+     * @param value name of the profile to use, {@code null} is ignored
+     * @return This resolver for fluent API usage
+     */
+    ConfigValueResolver withProfile(String value);
+
+    /**
+     * Provides a raw property value default value.
+     * <p>
+     * The raw default is considered to be on the source level and therefore takes precedence over typed defaults
+     * provided to any of the {@code as}-methods. Should raw default be empty of fail to convert to the target type the
+     * typed default is used.
+     *
+     * @param value raw default value, not {@code null}
+     * @return This resolver for fluent API usage
+     */
+    ConfigValueResolver withDefault(String value);
+
+    /**
+     * Disables the default behaviour of considering properties defined as empty string as being not present (missing).
+     * Effectively this means {@code as}-methods might then return empty things in case the property is defined as empty
+     * in the source.
+     *
+     * @return This resolver for fluent API usage
+     */
+    ConfigValueResolver acceptEmpty();
+
+    /**
+     * Disables the default behaviour of not throwing exceptions and instead returning default values for case of
+     * missing or empty raw value. If used in combination with {@link #acceptEmpty()} empty raw values will not throw an
+     * exception, otherwise they do.
+     *
+     * @return This resolver for fluent API usage
+     */
+    ConfigValueResolver throwOnMissingProperty();
```

Comment:
If a configuration value is an empty string, it is because a user has deliberately deleted that property value, explicitly overriding the default value.  See #531 for the use case.  It's not a question of throwing or not throwing.  Empty values should always be represented by `Optional*.empty()` which works for any value type.